Fix: #868 -Wsign-compare

## Maintainer Summary

2026-04-24 01:56:42 UTC

### Post Merge Audit

#868 

Usual process: For every Merge, the Assigned Maintainer is responsible for addressing any errors & warnings as a result of the Merge and applying the label Merge Defect. After Opening an Issue and performing Triage. the Maintainer is then responsible to cure the Merge Defect via PR.

```
/home/h02332/head/iccDEV/IccProfLib/IccIO.cpp:616:14: warning: comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'int64_t' (aka 'long') [-Wsign-compare]
  616 |     if (nPos < m_nStartPos)
      |         ~~~~ ^ ~~~~~~~~~~~
/home/h02332/head/iccDEV/IccProfLib/IccIO.cpp:616:14: warning: comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'int64_t' (aka 'long') [-Wsign-compare]
  616 |     if (nPos < m_nStartPos)
      |         ~~~~ ^ ~~~~~~~~~~~
```

--- a/.github/ci/regression/embedio-read8-bounds.cpp
+++ b/.github/ci/regression/embedio-read8-bounds.cpp
@@ -1,0 +1,56 @@
+#include "IccIO.h"
+
+#include <cstdio>
+
+int main()
+{
+  icUInt8Number bytes[16];
+  for (size_t i = 0; i < sizeof(bytes); ++i) {
+    bytes[i] = (icUInt8Number)('A' + i);
+  }
+
+  CIccMemIO backing;
+  if (!backing.Attach(bytes, sizeof(bytes))) {
+    std::printf("backing attach failed\n");
+    return 1;
+  }
+
+  if (backing.Seek(8, icSeekSet) != 8) {
+    std::printf("initial seek failed\n");
+    return 1;
+  }
+
+  CIccEmbedIO embed;
+  if (!embed.Attach(&backing, 4)) {
+    std::printf("embed attach failed\n");
+    return 1;
+  }
+
+  icUInt8Number out[16] = {};
+
+  if (backing.Seek(0, icSeekSet) != 0) {
+    std::printf("backing rewind failed\n");
+    return 1;
+  }
+
+  size_t read = embed.Read8(out, sizeof(out));
+  if (read != 0) {
+    std::printf("Read8 read before embedded window: read=%zu\n", read);
+    return 1;
+  }
+
+  if (embed.Seek(2, icSeekSet) != 2) {
+    std::printf("embed seek failed\n");
+    return 1;
+  }
+
+  read = embed.Read8(out, sizeof(out));
+  if (read != 2 || out[0] != 'K' || out[1] != 'L') {
+    std::printf("Read8 did not clamp to embedded window: read=%zu first=%u second=%u\n",
+                read, (unsigned)out[0], (unsigned)out[1]);
+    return 1;
+  }
+
+  std::printf("CIccEmbedIO::Read8 bounds OK\n");
+  return 0;
+}

--- a/.github/instructions/testing.instructions.md
+++ b/.github/instructions/testing.instructions.md
@@ -37,8 +37,9 @@ ctest --test-dir out/vs2022-x64 -C Release --output-on-failure --no-tests=error
 cmake --build out/vs2022-x64 --config Release --target check
 ```
 
-Linux currently registers 26 CTest suites. Windows full tool builds currently
-register 6 CTest suites: the IccConnect threaded CMM regression, two
+Linux currently registers 27 CTest suites. Windows full tool builds currently
+register 7 CTest suites: the IccConnect threaded CMM regression, the EmbedIO
+Read8 bounds regression, two
 batch-backed suites through `Build/Cmake/Testing/RunWindowsBatchTest.cmake`, the
 iccDumpProfile smoke suite, the issue-987 shared export suite, and the PAWG
 report smoke suite. Windows feature-disabled builds register the subset whose

--- a/.github/skills/maintainer-ci-ctest/SKILL.md
+++ b/.github/skills/maintainer-ci-ctest/SKILL.md
@@ -45,7 +45,7 @@ when practical.
 
 - `check` must exist on every platform.
 - `check` and workflow CTest execution must use `--no-tests=error`.
-- Linux suite count assertions currently expect `Total Tests: 26`.
+- Linux suite count assertions currently expect `Total Tests: 27`.
 - Before committing a CTest suite-count change, run
   `rg "Total Tests:|Linux currently registers|Linux suite count assertions" .github docs`
   and update every stale count in the same change.

--- a/.github/workflows/ci-iccdev-tool-tests.yml
+++ b/.github/workflows/ci-iccdev-tool-tests.yml
@@ -207,7 +207,7 @@ jobs:
           git config --global credential.helper ""
           unset GITHUB_TOKEN || true
 
-          EXPECTED_CTEST_TOTAL=26
+          EXPECTED_CTEST_TOTAL=27
           CTEST_LIST="${BUILD_DIR}/ctest-list.txt"
           ctest --test-dir "${BUILD_DIR}" -N --no-tests=error | tee "${CTEST_LIST}"
           if ! grep -q "^Total Tests: ${EXPECTED_CTEST_TOTAL}$" "${CTEST_LIST}"; then

--- a/Build/Cmake/Testing/CMakeLists.txt
+++ b/Build/Cmake/Testing/CMakeLists.txt
@@ -117,6 +117,41 @@ function(iccdev_add_iccconnect_thread_test)
   endif()
 endfunction()
 
+function(iccdev_add_embedio_read8_bounds_test)
+  if(NOT TARGET "${TARGET_LIB_ICCPROFLIB}")
+    return()
+  endif()
+
+  add_executable(iccEmbedIoRead8BoundsTest
+    "${ICCDEV_REPO_ROOT}/.github/ci/regression/embedio-read8-bounds.cpp"
+  )
+  target_link_libraries(iccEmbedIoRead8BoundsTest PRIVATE ${TARGET_LIB_ICCPROFLIB})
+  add_dependencies(check iccEmbedIoRead8BoundsTest)
+
+  add_test(
+    NAME iccdev.embedio-read8-bounds
+    COMMAND "$<TARGET_FILE:iccEmbedIoRead8BoundsTest>"
+  )
+  set_tests_properties(iccdev.embedio-read8-bounds PROPERTIES
+    WORKING_DIRECTORY "${ICCDEV_REPO_ROOT}"
+    TIMEOUT 60
+    LABELS "iccdev;embedio;regression"
+  )
+  if(WIN32)
+    set(_embedio_windows_env_mods
+      "PATH=path_list_prepend:$<TARGET_FILE_DIR:iccEmbedIoRead8BoundsTest>"
+      "PATH=path_list_prepend:$<TARGET_FILE_DIR:${TARGET_LIB_ICCPROFLIB}>"
+    )
+    foreach(_runtime_path IN LISTS ICCDEV_WINDOWS_RUNTIME_PATHS)
+      list(APPEND _embedio_windows_env_mods
+        "PATH=path_list_prepend:${_runtime_path}")
+    endforeach()
+    set_tests_properties(iccdev.embedio-read8-bounds PROPERTIES
+      ENVIRONMENT_MODIFICATION "${_embedio_windows_env_mods}"
+    )
+  endif()
+endfunction()
+
 if(WIN32)
   function(iccdev_add_windows_batch_test NAME TIMEOUT LABELS SCRIPT)
     set(_one_value_args EXPECTED_PROFILE_PARSE_COUNT EXPECTED_OUTPUT FORBIDDEN_REGEX)
@@ -265,6 +300,7 @@ if(WIN32)
   endif()
 
   iccdev_add_iccconnect_thread_test()
+  iccdev_add_embedio_read8_bounds_test()
 
   return()
 endif()
@@ -315,6 +351,7 @@ set(ICCDEV_TEST_ENV
 )
 
 iccdev_add_iccconnect_thread_test()
+iccdev_add_embedio_read8_bounds_test()
 
 function(iccdev_add_script_test NAME TIMEOUT LABELS WORKING_DIRECTORY SCRIPT)
   if(NOT EXISTS "${SCRIPT}")

--- a/IccProfLib/IccIO.cpp
+++ b/IccProfLib/IccIO.cpp
@@ -621,10 +621,15 @@ size_t CIccEmbedIO::Read8(void *pBuf, size_t nNum)
     return 0;
 
   if (m_nSize > 0) {
-    size_t nPos = m_pIO->Tell();
-    size_t nOffset = nPos - m_nStartPos;
+    int64_t nPos = m_pIO->Tell();
+    if (nPos < m_nStartPos)
+      return 0;
 
-    if (nOffset + nNum > m_nSize)
+    size_t nOffset = (size_t)(nPos - m_nStartPos);
+    if (nOffset > m_nSize)
+      return 0;
+
+    if (nNum > m_nSize - nOffset)
       nNum = m_nSize - nOffset;
   }
 

--- a/docs/ctest.md
+++ b/docs/ctest.md
@@ -90,11 +90,12 @@ cannot pass as a green no-op.
 
 ## Registered Suites
 
-Linux currently registers 26 tests:
+Linux currently registers 27 tests:
 
 | Test | Source |
 |------|--------|
 | `iccdev.create-profiles` | `Testing/CreateAllProfiles.sh` |
+| `iccdev.embedio-read8-bounds` | `.github/ci/regression/embedio-read8-bounds.cpp` |
 | `iccdev.iccconnect-threaded-cmm` | `.github/ci/regression/iccconnect-threaded-cmm.cpp` |
 | `iccdev.legacy-run-tests` | `Testing/RunTests.sh` |
 | `iccdev.tool-coverage` | `.github/scripts/iccdev-tool-coverage-baseline.sh --asan --skip-hybrid` |
@@ -162,10 +163,11 @@ without signed-conversion sanitizer findings.
 and fails if `iccFromXml` emits sanitizer diagnostics from string-size
 arithmetic.
 
-Windows full tool builds currently register 6 tests:
+Windows full tool builds currently register 7 tests:
 
 | Test | Source |
 |------|--------|
+| `iccdev.embedio-read8-bounds` | `.github/ci/regression/embedio-read8-bounds.cpp` |
 | `iccdev.iccconnect-threaded-cmm` | `.github/ci/regression/iccconnect-threaded-cmm.cpp` |
 | `iccdev.windows-create-profiles` | `Testing/CreateAllProfiles.bat` |
 | `iccdev.windows-legacy-run-tests` | `Testing/RunTests.bat` |

